exporter/prometheusremotewriteexporter/wal.go

```diff
@@ -139,13 +156,136 @@ func (prwe *prweWAL) start(ctx context.Context) error {
 	return nil
 }
 
+func (prwe *prweWAL) run(ctx context.Context) (err error) {
+	if err := prwe.start(ctx); err != nil {
+		return err
+	}
+
+	// Start the process of exporting but wait until the exporting has started.
+	waitUntilStartedCh := make(chan bool)
+	go func() {
+		prwe.continuallyPopWALThenExport(ctx, func() { close(waitUntilStartedCh) })
+	}()
+	<-waitUntilStartedCh
+	return nil
+}
+
+// continuallyPopWALThenExport reads a prompb.WriteRequest proto encoded blob from the WAL, and moves
+// the WAL's front index forward until either the read buffer period expires or the maximum
+// buffer size is exceeded. When either of the two conditions are matched, it then exports
+// the requests to the Remote-Write endpoint, and then truncates the head of the WAL to where
+// it last read from.
+func (prwe *prweWAL) continuallyPopWALThenExport(ctx context.Context, signalStart func()) (err error) {
+	var reqL []*prompb.WriteRequest
+	defer func() {
+		// Keeping it within a closure to ensure that the later
+		// updated value of reqL is always flushed to disk.
+		if errL := prwe.exportSink(ctx, reqL); len(errL) != 0 && err == nil {
+			err = multierror.Append(nil, errL...)
+		}
+	}()
+
+	freshTimer := func() *time.Timer {
+		return time.NewTimer(prwe.walConfig.truncateFrequency())
+	}
+
+	timer := freshTimer()
+	defer func() {
+		// Added in a closure to ensure we capture the later
+		// updated value of timer when changed in the loop below.
+		timer.Stop()
+	}()
+
+	signalStart()
+
+	maxCountPerUpload := prwe.walConfig.bufferSize()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-prwe.stopChan:
+			return nil
+		default:
+		}
+
+		req, err := prwe.readPrompbFromWAL(ctx, atomic.LoadUint64(&prwe.rWALIndex))
+		if err != nil {
+			return err
+		}
+		reqL = append(reqL, req)
+
+		// Now increment the WAL's read index.
+		atomic.AddUint64(&prwe.rWALIndex, 1)
+
+		shouldExport := false
+		select {
+		case <-timer.C:
+			shouldExport = true
+		default:
+			shouldExport = len(reqL) >= maxCountPerUpload
```

Comment:
When will this be called? Won't this `select` always fall into the `timer.C`?

Reply:
Most often I'd expect this to hit the `default` case, actually.  Unless the timer is very short I would expect multiple metrics to be processed from the WAL between timer expirations.  The default timer length is 1min and `readPrompbFromWAL()` looks like it shouldn't block for more than about 8s at the outside.